src/ray/gcs/gcs_server/gcs_resource_manager.h

```diff
@@ -161,17 +169,28 @@ class GcsResourceManager : public rpc::NodeResourceInfoHandler {
   /// Send any buffered resource usage as a single publish.
   void SendBatchedResourceUsage();
 
+  /// Prelocked version of GetResourceUsageBatchForBroadcast. This is necessary for need
+  /// the functionality as part of a larger transaction.
+  void GetResourceUsageBatchForBroadcast_Locked(rpc::ResourceUsageBatchData &buffer)
+      EXCLUSIVE_LOCKS_REQUIRED(resource_buffer_mutex_);
+
   /// The runner to run function periodically.
   PeriodicalRunner periodical_runner_;
   /// Newest resource usage of all nodes.
   absl::flat_hash_map<NodeID, rpc::ResourcesData> node_resource_usages_;
-  /// A buffer containing resource usage received from node managers in the last tick.
-  absl::flat_hash_map<NodeID, rpc::ResourcesData> resources_buffer_;
+
+  /// Protect the lightweight heartbeat deltas which are accessed by different threads.
+  absl::Mutex resource_buffer_mutex_;
+  /// A buffer containing the lightweight heartbeats since the last broadcast.
+  absl::flat_hash_map<NodeID, rpc::ResourcesData> resources_buffer_
+      GUARDED_BY(resource_buffer_mutex_);
 
   /// A publisher for publishing gcs messages.
   std::shared_ptr<gcs::GcsPubSub> gcs_pub_sub_;
   /// Storage for GCS tables.
   std::shared_ptr<gcs::GcsTableStorage> gcs_table_storage_;
+  /// Whether or not to broadcast resource usage via redis.
+  const bool broadcast_enabled_;
```

Comment:
```suggestion
  const bool redis_broadcast_enabled_;
```

Consider always distinguishing between grpc/redis in the variable and method names to avoid ambiguity.